### PR TITLE
setup.py: add commands test&plugin, add a CI for setup.py and docs update

### PR DESCRIPTION
**Pull request by ana**
This is an intent to wrap up https://github.com/avocado-framework/avocado/issues/4372 and also includes a v2 of one of the commits from https://github.com/avocado-framework/avocado/pull/4798 (command plugin); I have chosen to leave the `Makefile` updates for a different PR

I'm adding two new commands to `setup.py`: plugin and selftests. A new CI pipeline to test well all the commands provided by `setup.py` and updating the documentation.
The new CI pipeline can be seen at https://github.com/ana/avocado/actions/runs/1090390582

### CODE REVIEW

File: docs/source/guides/contributor/chapters/how.rst [hunk 1]
```diff
@@ -83,7 +83,7 @@ Git workflow
   manual.
 
 - Make sure your code is working (install your version of avocado, test
-  your change, run ``make check`` to make sure you didn't introduce any
+  your change, run ``python3 setup.py selftests`` to make sure you didn't introduce any
```
Comment (richtja) [inline docs/source/guides/contributor/chapters/how.rst:86]: I think that for now, the `python3 setup.py selftests` is not a substitution for `make check` because the `make check` will restart the environment  (it runs make clean uninstall develop) and then runs selftests.

File: setup.py [hunk 2]
```diff
@@ -203,6 +204,56 @@ def run(self):
             sys.exit(128)
 
 
+class Selftests(SimpleCommand):
+    """Run selftests"""
+
+    description = 'Run selftests'
+    user_options = [
+        ("job-api", None, "Run job API checks"),
+        ("static-checks", None, "Run static checks (isort, lint, etc)"),
```
Comment (beraldoleal) [inline setup.py:213]: job-api and static-checks are also inside "selftests" category. Any reason for not using the same prefix here?

File: setup.py [hunk 3]
```diff
@@ -203,6 +204,56 @@ def run(self):
...
+        self.job_api = False  # pylint: disable=W0201
+        self.static_checks = False  # pylint: disable=W0201
+        self.selftests_nrunner_interface = False  # pylint: disable=W0201
+        self.selftests_unit = False  # pylint: disable=W0201
+        self.selftests_jobs = False  # pylint: disable=W0201
+        self.selftests_functional = False  # pylint: disable=W0201
+        self.selftests_optional_plugins = False  # pylint: disable=W0201
+        self.disable_plugin_checks = []  # pylint: disable=W0201
+        self.features = False  # pylint: disable=W0201
+        self.all = False  # pylint: disable=W0201
```
Comment (beraldoleal) [inline setup.py:234]: I'm wondering if this shouldn't be `True` by default. Any reason for that?

File: setup.py [hunk 4]
```diff
@@ -356,5 +407,6 @@ def run(self):
           cmdclass={'clean': Clean,
                     'develop': Develop,
                     'lint': Linter,
```
Comment (beraldoleal) [inline setup.py:446]: Will this `lint` target be away, since we have the `static-checks` now?

File: setup.py [hunk 5]
```diff
@@ -274,6 +274,43 @@ def run(self):
...
+            plugins_list.append(plugin.parts[-2])
+
+        if self.list or (not self.list and not self.install):
+            print("List of available plugins:\n ", "\n  ".join(plugins_list))
+            return
+
+        if self.install in plugins_list:
+            action = ["install"]
+            if self.user:
+                action += ["--user"]
```
Comment (beraldoleal) [inline setup.py:306]: Thanks! :)

File: .github/workflows/setup.yml [hunk 6]
```diff
@@ -0,0 +1,234 @@
+name: Setup tests
```
Comment (beraldoleal) [inline .github/workflows/setup.yml:1]: Very important those tests. Really nice!

My only concern here is that this probably will be missed before the release, unless we decide to run the release always after Mondays. Is there any way to schedule this, also when any milestone is closed?

**Review by beraldoleal — CHANGES_REQUESTED** [review]
Hi @ana , thank you for this. Very welcome.

Just left a few comments for you.

File: docs/source/guides/contributor/chapters/how.rst (hunk 1 above)
Comment (ana) [reply docs/source/guides/contributor/chapters/how.rst:86]: @richtja good point. I'm going to re-work to mention quick the quick tests with `python3 setup.py lint` and then `make check` for the longer tests. 
I have a small annoyance with `make check` because I like more `make selftests` but this is not the PR for discussing that :D

File: setup.py (hunk 2 above)
Comment (ana) [reply setup.py:213]: The options naming are inherited from `selftests/check.py`, see https://github.com/avocado-framework/avocado/blob/master/selftests/check.py#L198

I fully agree about using the same prefix to all of them. In this case, given the context, isn't maybe better just remove the "selftest" from all of them?  They're always used  in a context that makes clear is a selftests, either with
`python3 setup.py selftest --static-checks`
or with:
`./selftests/check.py --static-checks`

File: setup.py (hunk 4 above)
Comment (ana) [reply setup.py:446]: This is something we need to discuss. There is still a difference between both options, `python3 setup.py lint` works without installing avocado while `python3 setup.py selftests --static-checks` needs avocado installed (in develop mode) and uses avocado to run the tests.

File: .github/workflows/setup.yml (hunk 6 above)
Comment (ana) [reply .github/workflows/setup.yml:1]: We can schedule them another day  or run them manually.  Eventually we should be able to add them to the (semi?)automatic tests for the release.

File: setup.py (hunk 3 above)
Comment (ana) [reply setup.py:234]: > I'm wondering if this shouldn't be `True` by default. Any reason for that?

Yes, because by default this option isn't selected by the user unless they type `--all`. To be honest, the the option is cosmetic right now, both `python3 setup.py selftests` and `python3 setup.py selftests --all` will run all the tests. But I think eventually we might default to something running only some tests if nothing is provided instead of running all of them. Thoughts?

**Comment by ana** [comment]
@beraldoleal @richtja Thank you for your review and comments. I had some questions  above before a v2 and I'd love for your input =)

File: setup.py (hunk 2 above)
Comment (beraldoleal) [reply setup.py:213]: Sure, lets remove. I agree.

File: setup.py (hunk 3 above)
Comment (beraldoleal) [reply setup.py:234]: If `python3 setup.py selftests` will be the same as `python3 setup.py selftests --all` than the default for `--all` should be true, right? And maybe we don't need `--all` at all.

Would be nice to avoid having multiple commands to run all our tests.

File: setup.py (hunk 4 above)
Comment (beraldoleal) [reply setup.py:446]: I see. Well I don't see a reason to have both here. And since avocado is already used by others checks, I'm not sure if we need the first one. And TBH, unless we have a strong argument, having mixing checks were some uses avocado and others don´t, seems confusing. So let's see what others think about this.

File: .github/workflows/setup.yml (hunk 6 above)
Comment (beraldoleal) [reply .github/workflows/setup.yml:1]: Thinking out of loud: We do have a target that is when a milestone is closed, that is something we have to do anyway. I would prefer this over running manually.

Comment (richtja) [reply .github/workflows/setup.yml:1]: We can use the milestone closing as a target, but we have to define when the milestone will be closed. Right now is up to the facilitator, sometime is before the release meeting, sometimes is after the release meeting and sometimes is during the meeting and the same is with the release itself and that can caused problems with tests. So IMO if we will be using this target we have to put it in the pre-release checks.

File: setup.py (hunk 4 above)
Comment (richtja) [reply setup.py:446]: IMO the selftest checks is for the contributors whose have to work in develop mode anyway, so I agree with @beraldoleal  that we do not need the `python3 setup.py lint` check

**Comment by ana** [comment]
Thanks for the comments, I have forced-pushed an update with the following changes:

* Reverted usage of `python3 setup.py selftests`  at docs/source/guides/contributor/chapters/how.rst

* Renamed  selftests options removing `--selftstes` prefixes

* Remove cosmetic option `--all` from `python3 setup.py selftests`

* Deprecate `lint` command in favour of `selftests --static-checks`, updating GH Actions.

I haven't addressed the part about running setup.yml in releases.

**Comment by ana** [comment]
Added new commit to update the packaging spec. Maybe it's better to run all the selftests directly, including the static checks.

File: selftests/check.py [hunk 7]
```diff
@@ -566,9 +574,32 @@ def print_failed_tests(tests):
...
+    args.unit = True
+    args.jobs = True
+    args.functional = True
+    args.optional_plugins = True
+
+
+def main(args):  # pylint: disable=W0621
+
+    # pylint: disable=R0916
+    if (not args.static_checks and not args.job_api and
```
Comment (clebergnu) [inline selftests/check.py:590]: I'd replace this long conditional with:

```python
if not any([args.static_checks, args.job_api, ...]):
```

File: selftests/check.py [hunk 8]
```diff
@@ -604,4 +635,5 @@ def main():
 
 
 if __name__ == '__main__':
-    sys.exit(main())
+    args = parse_args()
```
Comment (clebergnu) [inline selftests/check.py:638]: I'm curious: any reason for this change?

File: setup.py (hunk 4 above)
Comment (clebergnu) [reply setup.py:432]: @ana have you considered using the `test` command instead?  I mean, Python developers will expect to be able to run `python setup.py test`, so replacing it, instead of adding `selftests` seems like the best option here.

File: docs/source/guides/user/chapters/installing.rst [hunk 9]
```diff
@@ -119,13 +119,25 @@ Installing from source code
...
 Then to install Avocado from the git repository run::
 
     $ git clone git://github.com/avocado-framework/avocado.git
     $ cd avocado
-    $ sudo python3 setup.py install
+    $ python3 setup.py install --user
+
+Optionally, to install the plugins run:
+
+    $ python3 setup.py plugin --install=golang --user
```
Comment (clebergnu) [inline docs/source/guides/user/chapters/installing.rst:132]: I think we should have a generated page per plugin on our docs, with an index and refer to them here.  The idea is to avoid a maintenance burden and having this list out of sync with the existing plugins.  May be done in a separate PR.

File: python-avocado.spec [hunk 10]
```diff
@@ -174,7 +174,7 @@ PATH=%{buildroot}%{_bindir}:%{buildroot}%{_libexecdir}/avocado:$PATH \
     PYTHONPATH=%{buildroot}%{python3_sitelib}:. \
     LANG=en_US.UTF-8 \
     AVOCADO_CHECK_LEVEL=0 \
-    %{python3} selftests/check.py --disable-static-checks --disable-plugin-checks=robot
+    %{python3} selftests/check.py --job-api --nrunner-interface --unit --jobs --functional --optional-plugins --disable-plugin-checks=robot
```
Comment (clebergnu) [inline python-avocado.spec:177]: Every change to the `python-avocado.spec` file should bump the *release* number and add a `Changelog` entry.

File: .cirrus.yml [hunk 11]
```diff
@@ -71,7 +71,7 @@ debian_egg_task:
 fedora_selftests_task:
     selftests_script:
        - make develop
-       - PATH=$HOME/.local/bin:$PATH LANG=en_US.UTF-8 AVOCADO_CHECK_LEVEL=0 python3 selftests/check.py --disable-static-checks
+       - PATH=$HOME/.local/bin:$PATH LANG=en_US.UTF-8 AVOCADO_CHECK_LEVEL=0 python3 selftests/check.py
```
Comment (clebergnu) [inline .cirrus.yml:74]: This is causing failures in the Cirrus CI jobs.  They are run on an environment as close as possible to the RPM build environment, and they lack the lint tools.  Given the logic was inverted from "include" to "exclude", we have to include all but the `--static-check` options here.

**Review by clebergnu — CHANGES_REQUESTED** [review]
Hi @ana, this is great work! Besides the requests for change in some comments, I'm curious if  you plan to sync this somehow into avocado-vt.

File: setup.py (hunk 4 above)
Comment (ana) [reply setup.py:432]: > @ana have you considered using the `test` command instead? I mean, Python developers will expect to be able to run `python setup.py test`, so replacing it, instead of adding `selftests` seems like the best option here.

yes, I discarded it because `python setup.py test` already does something (running unit test).
The command is deprecated:  you get a message **"WARNING: Testing via this command is deprecated and will be removed in a future version. Users looking for a generic test entry point independent of test runner are encouraged to use tox"**

If you prefer it, I'm happy to override the actual command and use `test ` instead of `selftests`

Comment (clebergnu) [reply setup.py:432]: > > @ana have you considered using the `test` command instead? I mean, Python developers will expect to be able to run `python setup.py test`, so replacing it, instead of adding `selftests` seems like the best option here.
> 
> yes, I discarded it because `python setup.py test` already does something (running unit test).
> The command is deprecated: you get a message **"WARNING: Testing via this command is deprecated and will be removed in a future version. Users looking for a generic test entry point independent of test runner are encouraged to use tox"**
> 

OK... still, it's not clear to me if they are deprecating the *built-in runner* or the command/interface.

> If you prefer it, I'm happy to override the actual command and use `test ` instead of `selftests`

I guess my final preference will depend on the previous point... do you think we can find that out?

File: selftests/check.py (hunk 8 above)
Comment (ana) [reply selftests/check.py:638]: Yes, if you look at the remaining changes, there is a call to `selftests.check.main(args)` from `setup.py`
Please, tell me if you know of a better way to do this.

Comment (clebergnu) [reply selftests/check.py:638]: > Yes, if you look at the remaining changes, there is a call to `selftests.check.main(args)` from `setup.py`

Oh, I missed this.  Thanks!

> Please, tell me if you know of a better way to do this.

Nope, that's fine!

File: setup.py (hunk 4 above)
Comment (ana) [reply setup.py:432]: > OK... still, it's not clear to me if they are deprecating the _built-in runner_ or the command/interface.
> 
> > If you prefer it, I'm happy to override the actual command and use `test ` instead of `selftests`
> 
> I guess my final preference will depend on the previous point... do you think we can find that out?

The setuptools command is being deprecated. More information at https://github.com/pypa/setuptools/issues/1684

File: .github/workflows/setup.yml (hunk 6 above)
Comment (ana) [reply .github/workflows/setup.yml:1]: @beraldoleal @richtja Thanks for the comments. IMHO this is something we should discuss at https://github.com/avocado-framework/avocado/issues/4692

File: docs/source/guides/user/chapters/installing.rst (hunk 9 above)
Comment (ana) [reply docs/source/guides/user/chapters/installing.rst:132]: > I think we should have a generated page per plugin on our docs, with an index and refer to them here. The idea is to avoid a maintenance burden and having this list out of sync with the existing plugins. May be done in a separate PR.

Ok, do you mind filing a bug with the idea you have in mind?

**Comment by ana** [comment]
> Hi @ana, this is great work! Besides the requests for change in some comments, I'm curious if you plan to sync this somehow into avocado-vt.

Thanks for the review @clebergnu. I'll do eventually the sync with avocado-vt, I have filed a bug  https://github.com/avocado-framework/avocado-vt/issues/3167

I just force-pushed an update on this, changes:
* cirrus.yml: fedora_selftests: update to use new options of check.py
* update changelog in spec file
* use  if not any([...]) for long conditional
* rename command `selftest` to `test` command (it'll override the existing one), update `setup.yml `accordingly

File: setup.py [hunk 12]
```diff
@@ -203,6 +204,51 @@ def run(self):
...
 
 
+class Test(SimpleCommand):
+    """Run selftests"""
+
+    description = 'Run selftests'
+    user_options = [
+        ("job-api", None, "Run job API checks"),
+        ("static-checks", None, "Run static checks (isort, lint, etc)"),
+        ("nrunner-interface", None, "Run selftests/functional/test_nrunner_interface.py"),
```
Comment (beraldoleal) [inline setup.py:205]: nrunner-interface is a specific type of "functional". While I understand the importance of this test, I'm wondering what was the criteria to create a specific target for it here.

My concern is that in the future this list will be huge because every single functional/unit test will have an own user-option. What do you think?

File: setup.py [hunk 13]
```diff
@@ -203,6 +204,51 @@ def run(self):
...
+        ("unit", None, "Run selftests/unit/"),
+        ("jobs", None, "Run selftests/jobs/"),
+        ("functional", None, "Run selftests/functional/"),
+        ("optional-plugins", None, "Run optional_plugins/*/tests/"),
+        ("disable-plugin-checks", None, "Disable checks for a plugin (by directory name)"),
+        ("features", None, "Show the features tested by this test")
```
Comment (beraldoleal) [inline setup.py:211]: nitpick: maybe s/features/list-features/g or s/features/show-features/g

File: setup.py (hunk 4 above)
Comment (beraldoleal) [reply setup.py:432]: IIUC, we mixed the topics here. My initial comment it was about 'lint' command. And by @richtja comment and @clebergnu thumbs up, I'm assuming that we could drop it. Did I understood it right?

File: .github/workflows/setup.yml (hunk 6 above)
Comment (beraldoleal) [reply .github/workflows/setup.yml:1]: Hi @ana, my understand is that yes, this is also part of #4692, but there we are more concerned about the package creation/distribution itself.

So I'm not so sure about it. You are introducing a new test workflow here, and it is important to see the "best" schedule timing for it. Otherwise we are going to have the impression that all is covered, but in fact, any last minute change can break the installation because of timing. And since we usually fix things during the feature freeze at the end of week, this "timing" might be an issue.

While we could adopt what @richtja suggested adding a new item to our pre-release check, currently we are also trying to reduce the amount of manual checks before the release (#3456).

File: docs/source/guides/user/chapters/installing.rst (hunk 9 above)
Comment (beraldoleal) [reply docs/source/guides/user/chapters/installing.rst:132]: Another idea is to have each plugin with its own README.rst updated (since some people might reach a plugin by the github page directly) and try to reuse this README inside our docs somehow without content duplication, so when a plugin'sREADME is updated our documentation will be also updated.

File: setup.py [hunk 14]
```diff
@@ -186,22 +186,13 @@ def finalize_options(self):
...
-            run(['selftests/isort.sh'], check=True)
-            run(['selftests/lint.sh'], check=True)
-            run(['selftests/signedoff-check.sh'], check=True)
-            run(['selftests/spell.sh'], check=True)
-            run(['selftests/modules-boundaries.sh'], check=True)
-        except CalledProcessError as e:
-            print("Failed during lint checks: ", e)
-            sys.exit(128)
+        print("This command is deprecated, please use instead: python3 setup.py selftests --static-checks")
+        sys.exit()
```
Comment (beraldoleal) [inline setup.py:195]: Ok, Here is the removal that I was looking for. :) So, please ignore my previous comment about lint.

And unless someone has a strong argument about keeping the execution working (only adding the warning message for one or two releases), I'm ok with this approach.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -34,8 +34,6 @@ jobs:
         run: avocado --version
```
Comment (beraldoleal) [inline .github/workflows/ci.yml:34]: I wouldn't mind if this commit gets squashed with the last one, since the last one it is breaking this workflow.

File: .cirrus.yml [hunk 16]
```diff
@@ -71,7 +71,7 @@ debian_egg_task:
 fedora_selftests_task:
     selftests_script:
        - make develop
-       - PATH=$HOME/.local/bin:$PATH LANG=en_US.UTF-8 AVOCADO_CHECK_LEVEL=0 python3 selftests/check.py --disable-static-checks
```
Comment (beraldoleal) [inline .cirrus.yml:74]: This also could be merged with the commit that is removing the "disable*" commands.

**Review by beraldoleal — CHANGES_REQUESTED** [review]
Hi @ana thanks for the changes, I just left a few comments for you.

File: setup.py (hunk 4 above)
Comment (ana) [reply setup.py:432]: > IIUC, we mixed the topics here. My initial comment it was about 'lint' command. And by @richtja comment and @clebergnu thumbs up, I'm assuming that we could drop it. Did I understood it right?

Yeah, we discussed two different things. I addressed both:
- `lint` has been deprecated (instead of removing it, now it'll show a message telling you to use `python setup.py test --static-check`)
- `selftests` has been renamed to `test`.

File: .github/workflows/setup.yml (hunk 6 above)
Comment (ana) [reply .github/workflows/setup.yml:1]: @beraldoleal You're right, I liked the wrong issue 😅  I meant https://github.com/avocado-framework/avocado/issues/4601 that's indeed adding a new *automatic* pre-release.
I'm discussing this with @willianrampazzo on Monday, tell me if you want to join!

File: setup.py (hunk 12 above)
Comment (ana) [reply setup.py:205]: @beraldoleal I was following the logic we have in `check.py`. For some reason, there were a group of test name nrunner-interface and I followed that.

File: setup.py (hunk 13 above)
Comment (ana) [reply setup.py:211]: > nitpick: maybe s/features/list-features/g or s/features/show-features/g

Nope, this is also following `check.py`.

File: setup.py (hunk 14 above)
Comment (ana) [reply setup.py:195]: ... and ignore my replies above 👍🏻

File: setup.py (hunk 12 above)
Comment (beraldoleal) [reply setup.py:205]: I see @ana. 

@clebergnu, what do you think about my point? Do you have any objections on removing this?

File: setup.py (hunk 13 above)
Comment (beraldoleal) [reply setup.py:211]: IMO, we are creating a targets options to setup.py, I don't think we have to restricted to what is inside check.py.

File: setup.py [hunk 17]
```diff
@@ -186,22 +186,13 @@ def finalize_options(self):
...
-            run(['selftests/inspekt-style.sh'], check=True)
-            run(['selftests/isort.sh'], check=True)
-            run(['selftests/lint.sh'], check=True)
-            run(['selftests/signedoff-check.sh'], check=True)
-            run(['selftests/spell.sh'], check=True)
-            run(['selftests/modules-boundaries.sh'], check=True)
-        except CalledProcessError as e:
-            print("Failed during lint checks: ", e)
-            sys.exit(128)
+        print("This command is deprecated, please use instead: python3 setup.py selftests --static-checks")
```
Comment (clebergnu) [inline setup.py:194]: According to the command rename, this needs to become "python3 setup.py test --static-checks".

**Review by clebergnu — APPROVED** [review]
On my side, two things before approving:

1. the `s/selftests/test/` update in the error message (must)
2. the squashing of commits as suggested by @beraldoleal (highly recommended) 

Thanks a lot for tackling this @ana!

**Comment by beraldoleal** [comment]
> On my side, two things before approving:
> 
>     1. the `s/selftests/test/` update in the error message (must)
> 
>     2. the squashing of commits as suggested by @beraldoleal (highly recommended)
> 
> 
> Thanks a lot for tackling this @ana!

Could you just comment here too, please?

https://github.com/avocado-framework/avocado/pull/4817#discussion_r683633088

File: setup.py (hunk 12 above)
Comment (ana) [reply setup.py:205]: @beraldoleal I'm opening an issue with this, it risks to get buried here with so many things.

Comment (beraldoleal) [reply setup.py:205]: That is fine @ana. We can discuss this there or here, however IMO, the outcome of this discussion it is important for this PR.

Comment (beraldoleal) [reply setup.py:205]: Btw, probably you have notice, but I'm marking some threads here as "resolved" to get less "buried".

**Comment by ana** [comment]
@beraldoleal  @clebergnu  Thanks for the follow-ups.
I have forced-pushed a new version, changes:

* Squashed a few commits with related changes
* fix deprecation message from lint command s/selftests/test/

Besides, I have created an issue to discuss about nrunner-interface https://github.com/avocado-framework/avocado/issues/4828
@beraldoleal feel free to make a new change to rename the options for `python3 setup.py test`, I don't have a strong opinion in that.

File: setup.py (hunk 13 above)
Comment (beraldoleal) [reply setup.py:211]: Ok, I was revisiting this thread, take this as example.

I understand that here we passing the same arguments to `check.py`. But we also agree that `check.py` is not a good example, so IMO we don't have to replicate this model to the new `setup.py` targets, otherwise, when changing `check.py` latter we will need to change this too, creating a bad user experience.

My suggestion:

```python
mapping = {'list-features': 'features'}
...
args.features = mapping.get(self.features)
selftests.check.main(args)
```

So, in summary, let's think about good names without being influenced by `check,py` so we are free to change `check.py` latter without affecting users/developers.

Comment (clebergnu) [reply setup.py:211]: Given that the issue is not only about *listing* the features, but about selecting them, we'd need at the very least:

```setup.py test --list-features```

AND

```setup.py test --features=unit,functional,...```

Which is more flexible indeed, and allows for more maneuvering around `check.py` current state. But, it's also worse usability wise (users have to run `--list-features` first, and then with the valid options).

Having said that, I don't think it's worth implementing this logic of listing/enabling "features" into `setup.py test` itself. I believe its place is really in an Avocado job (`avocado.core.job.Job`) and with `features` meaning the `suites` of a job.

Comment (beraldoleal) [reply setup.py:211]: Even better... +1 for not having 'list/show' actions here.

Comment (clebergnu) [reply setup.py:211]: FIY, this is the issue about the ideas I listed above: https://github.com/avocado-framework/avocado/issues/4832

**Comment by ana** [comment]
I have pushed another update replacing `--features` with `--list-features` in both `setup.py` and` check.py`.  /cc @beraldoleal

**Review by beraldoleal — APPROVED** [review]

**Comment by beraldoleal** [comment]
@richtja are your comments solved here? Do we have a green light from you?

**Comment by richtja** [comment]
> @richtja are your comments solved here? Do we have a green light from you?

Yes, sorry, I forgot to give my approval.

**Review by richtja — APPROVED** [review]
